agent/setup.go

```diff
@@ -177,3 +183,116 @@ func registerWithGRPC(b grpcresolver.Builder) {
 	defer registerLock.Unlock()
 	grpcresolver.Register(b)
 }
+
+// getPrometheusDefs reaches into every slice of prometheus defs we've defined in each part of the agent, and appends
+//  all of our slices into one nice slice of definitions per metric type for the Consul agent to pass to go-metrics.
+func getPrometheusDefs() lib.PrometheusDefs {
+	serviceName := []string{"consul"}
+	var gauges = [][]prometheus.GaugeDefinition{
+		cache.Gauges,
+		consul.AutopilotGauges,
+		consul.RPCGauges,
+		consul.SessionGauges,
+		grpc.StatsGauges,
+		usagemetrics.Gauges,
+	}
+	var gaugeDefs []prometheus.GaugeDefinition
+	for _, g := range gauges {
+		// Set Consul to each definition's namespace
+		var withService []prometheus.GaugeDefinition
+		for _, gauge := range g {
+			gauge.Name = append(serviceName, gauge.Name...)
```

Comment:
I believe there are two problems here:

1. this assumes that the `TelemetryConfig.MetricsPrefix` will always be `consul`. That is the default value, but it can be changed by the user. I believe this will initialize the wrong metrics if a user specifies a different metrics prefix.
2. `append` does not modify the slice passed in as the first argument, but it can modify the underlying array used by that slice if the array is not at capacity. This is a subtle thing that can lead to very strange bugs. In this case the array backing `serviceName` could be used for every `gauge.Name`. So when the second iteration runs it will modify the value for the first iteration. Every gauge could end up with the same name, so only the very last one will get initialized. I believe it works currently because the array is being created with a capacity of 1, but to fix the first issue, that likely will not work anymore.

This is an example of the problem: https://play.golang.org/p/QFxtyGXrlqw

I believe the convention when using append is assign the result of append to the variable passed as the first argument. There are only a few rare cases when it can be assigned to another variable.

Reply:
Since `go-metrics` is the code that adds the prefix to the name it seems like it should also handle prepending this name. Otherwise every caller will have to guess at how `go-metrics` is building up the final metric name.

Reply:
Yes, this should absolutely be using `TelemetryConfig.MetricsPrefix` that is a bug.

With the slice prepends, I agree that these should be handled by go-metrics -- I think for this release it's ok if we handle them here so long as we write an issue for it. (I've added some todos as well) It would be impossible to assign the var back to the first var and still be performing a prepend operation, we'd be appending in that case. It looks like if we allocate a new literal for each prepend, we avoid any potential bugs. https://play.golang.org/p/F_zPCegmka7

Here's the prometheus page with a different `metrics_prefix` supplied and using the literals in the commit following this comment.

![Screen Shot 2020-11-16 at 1 54 28 PM](https://user-images.githubusercontent.com/938395/99312756-5924b680-2813-11eb-90b6-e5b0780221e2.png)